### PR TITLE
[OpenKAT v1.5] Multi XTDB and organizations management through organization code as slug

### PR DESCRIPTION
**Pull request by Rieven**
## Changes

- Every organization can be manages through one interface. 
- Switch to different organizations using the organization code in URL. 
- Each organization has its own: KAT-alogus, Octopoes/XTDB database connection, findings, tasks, etc..
- Different dashboard/crisis room for each organization. 
- Switch organizations in main menu
- Admins can manage organizations: add members, redteamer management
- Redteamer can switch to different organizations which they are member of and do the scans.
- Clients can still switch to organizations they are member of and see their details.

## Issue ticket number and link
https://github.com/minvws/nl-kat-rocky/pull/76
https://github.com/minvws/nl-rt-tim-abang/issues/825
https://github.com/minvws/nl-rt-tim-abang/issues/826

## Proof
_Please paste some screenshots or other proof of your (working) change here. If you feel that this is not required (e.g. this PR is trivial), note that here._

## Good to know
- Does this PR introduce or depend on API-incompatible changes? If yes: what do other users/developers need to do or confirm before merging?
  - NO
- Does this PR depend on a specific version of a library?
  - NO
- Does this PR require config, setup, or `.env` changes?
  - YES: XTDB relies on a multinode type: `XTDB_TYPE=xtdb-multinode`

## Checklist for author(s):
- [x] This PR comes from a `feature` or `hotfix` branch, in line with our git branching strategy;
- [ ] I have squashed all commits before merging this PR;
- [x] This PR is "bite-sized" and only focuses on a single issue, problem, or feature;
- [x] If a non-trivial PR: This PR is properly linked to an issue on the project board;
- [ ] If a non-trivial PR: I have added screenshots or some other proof that my code does what it is supposed to do;
- [x] I am not reinventing the wheel: there is no high-quality library that already has this feature;
- [x] I have changed the example `.env` files if I added, removed, or changed any config options, and I have informed others that they need to modify their `.env` files if required;
- [x] I have performed a self-review of my own code;
- [x] I have commented my code, particularly in hard-to-understand areas;
- [ ] I have made corresponding changes to the documentation, if necessary;
- [ ] I have written unit, integration, and end-to-end tests for the change that I made;

```
## Checklist for functional reviewer(s):
- [ ] If a non-trivial PR: This PR is properly linked to an issue on the project board;
- [ ] I have checked out this branch, and successfully ran `make kat`;
- [ ] I have ran `make test-rf` and all end-to-end Robot Framework tests pass;
- [ ] I confirmed that the PR's advertised `feature` or `hotfix` works as intended;
- [ ] I confirmed that there are no unintended functional regressions in this branch;

### What works:
* _bullet point + screenshot (if useful) per tested functionality_

### What doesn't work:
* _bullet point + screenshot (if useful) per tested functionality_

### Bug or feature?:
* _bullet point + screenshot (if useful) if it is unclear whether something is a bug or an intended feature._
```

```
## Checklist for code reviewer(s):
- [ ] The code passes the CI tests and linters;
- [ ] The code does not bypass authentication or security mechanisms;
- [ ] The code does not introduce any dependency on a library that has not been properly vetted;
- [ ] The code does not violate Model-View-Template and our other architectural principles;
- [ ] The code contains docstrings, comments, and documentation where needed;
- [ ] The code prioritizes readability over performance where appropriate;
- [ ] The code conforms to our agreed coding standards.
```

### CODE REVIEW

File: account/mixins.py [hunk 1]
```diff
@@ -1,29 +1,42 @@
...
-                    self.organizationmembers = OrganizationMember.objects.filter(organization=self.organization)
-                else:
-                    self.organizationmembers = OrganizationMember.objects.filter(
-                        user=self.request.user, organization=self.organization
-                    )
-                if not self.organizationmembers and not request.user.is_superuser:
-                    raise Http404()
-            except Organization.DoesNotExist:
-                raise Http404()
+        organization_code = kwargs["organization_code"]
```
Comment (underdarknl) [inline account/mixins.py:21]: Should be handle kwargs without an organization_code too? By using get() perhaps?

File: account/mixins.py [hunk 2]
```diff
@@ -1,29 +1,42 @@
...
-    organizationmembers = None
+def get_octopoes_api_connector(base_uri: str, organization_code: str):
+    return OctopoesAPIConnector(base_uri, organization_code)
+
+
+class OrganizationView(View):
+    def __init__(self, **kwargs):
+        super().__init__(**kwargs)
+        self.organization = None
+        self.octopoes_api_connector = None
```
Comment (underdarknl) [inline account/mixins.py:17]: Would it make sense to also setup the KAT-alogus client here?

File: katalogus/utils.py [hunk 3]
```diff
@@ -1,7 +1,8 @@
 from typing import Type, List
+
+from katalogus.client import get_katalogus, Plugin
 from octopoes.models import OOI
 from tools.models import Organization
-from katalogus.client import get_katalogus, Plugin
 
 
 def get_enabled_boefjes_for_ooi_class(ooi_class: Type[OOI], organization: Organization) -> List[Plugin]:
```
Comment (underdarknl) [inline katalogus/utils.py:8]: This reference to the organization would be more like the rest if it was available on an KAT-alogus client instance, like the octopoes client. Move to future refactoring imho.

File: account/mixins.py (hunk 2 above)
Comment (Rieven) [reply account/mixins.py:13]: I think it was a bad idea to put the api connector in this view and even more de KAT-alogus api connector. Not all views which has an organization needs a connector. The view which needs what it needs, calls it at the moment where it is needed.

**Comment by Darwinkel** [comment]
## Checklist for functional reviewer(s):
- [x] If a non-trivial PR: This PR is properly linked to an issue on the project board;
- [x] I have checked out this branch, and successfully ran `make kat`;
- [ ] I have ran `make test-rf` and all end-to-end Robot Framework tests pass;
- [ ] I confirmed that the PR's advertised `feature` or `hotfix` works as intended;
- [ ] I confirmed that there are no unintended functional regressions in this branch;

Tested with a clean `.env` and all branches set to `main`, except for Rocky.

### What works:
* Creating the `_dev` Development Organisation (no conflicts or bugs ;) )
* Creating a new organisation.
* Creating new members.
* Granting and revoking authorizations.
* Scheduling, running, and normalizing objects.
* Switching and interacting with different organisations (nice UX!).
* There do not appear to any regressions in existing functionality.

### What doesn't work:
* **Security issue**: Superadmins don't have permission to change clearance levels by default, but can still do so through the new multiselect option. Does this also apply to other users?
image

* It is unclear how, as an admin, I can authorize a newly created user. Granting Clearance Permission doesn't appear to help. This appears to apply to both organisations, regardless of roles.

image

* The button to switch Organisation looks odd, and does not work if the text is clicked (only works if the background is clicked).
image
image


* Creating an object for the secondary organisation gives an error (although the object is actually created). Attempting to access the object afterwards results in the same error.
image

* Accessing the superadmin's profile when in the context of the secondary organisation
image


### Bug or feature?:
* Superadmins don't have permission to change clearance levels by default - but the steps to grant these are not obvious from a user perspective. Would be nice if there was some info about how this is done + that it should be done for new installs.
* The navigation menu disappears when entering the Crisis Room. I understand that this is because the Crisis Room is not bound to a specific organisation, but it feels clumsy and unrefined from an UI/UX perspective that there is no context/info/hint provided.
* There is no UI/UX hint that one should accept OOI clearance through their profile (and that this has not been done yet).

File: account/mixins.py (hunk 2 above)
Comment (underdarknl) [reply account/mixins.py:13]: If the connector is cheap (eg, just an instance of an often used object) its fine to instantiate for every view. If it is expensive we could just make it a property which creates the connection once we first use it.
ideally, I think storing the current organization on the request object would make sense to keep it accessible/reachable everywhere without the need for plumping.

**Comment by Darwinkel** [comment]
## Checklist for functional reviewer(s):
- [x] If a non-trivial PR: This PR is properly linked to an issue on the project board;
- [x] I have checked out this branch, and successfully ran `make kat`;
- [ ] I have ran `make test-rf` and all end-to-end Robot Framework tests pass;
- [ ] I confirmed that the PR's advertised `feature` or `hotfix` works as intended;
- [ ] I confirmed that there are no unintended functional regressions in this branch;

Tested with a clean `.env` and all branches set to `main`, except for Rocky.

### What is fixed:
* The organisation button now looks good, regardless of window size or text length. Plus it is clickable! :)

### What is not fixed:
* A newly created redteamer (or other user; regardless of organisation) now receives a different error when logging in for the first time, but it is still not clear how an admin should proceed to verify the new account.
image


### What doesn't work:
* Changing the clearance of a newly created object in the primary and secondary organisation gives an error
image
This also happens when manually running a Boefje from the catalogus:
image


* As a superadmin with L4 clearance accepted, I cannot change clearance levels in bulk (multiselect permission issue?) for both the primary and secondary organisation
image

* As a superadmin with L4 clearance accepted, I cannot change the clearance level of a newly created object in the secondary organisation. Adding an indemnification to the organisation fixes this, but there is no UI/UX that indicates to an admin that one should do this after creating a new organisation.
image

**Comment by Darwinkel** [comment]
The most recent commits have fixed the clearance level in bulk / multiselect button, but changing the clearance now results in the MultipleObjectsReturned error.

**Review by Lisser — DISMISSED** [review]

**Review by Lisser — DISMISSED** [review]

**Review by Lisser — APPROVED** [review]

**Review by underdarknl — APPROVED** [review]

**Pull request merged**